common/src/main/java/com/microsoft/identity/common/internal/result/MsalBrokerResultAdapter.java

```diff
@@ -182,12 +183,112 @@ public BaseException getBaseExceptionFromBundle(@NonNull final Bundle resultBund
             return new BaseException(ErrorStrings.UNKNOWN_ERROR, "Broker Result not returned from Broker");
         }
 
-        BaseException baseException;
+        final String exceptionType = brokerResult.getExceptionType();
 
+        if(!TextUtils.isEmpty(exceptionType)){
+            return getBaseExceptionFromExceptionType(exceptionType, brokerResult);
+        }else {
+            // This code is here for legacy purposes where old versions of broker (3.1.8 or below)
+            // wouldn't return exception type in the result.
+            Logger.info(TAG, "Exception type is not returned from the broker, " +
+                    "using error codes to transform to the right exception");
+            return getBaseExceptionFromErrorCodes(brokerResult);
+        }
+
+    }
+
+    private BaseException getBaseExceptionFromExceptionType(@NonNull final String exceptionType,
+                                                            @NonNull final BrokerResult brokerResult){
+        BaseException baseException = null;
+
+        if(exceptionType.equalsIgnoreCase(UiRequiredException.sName)){
+            Logger.warn(TAG, "Received a UIRequired exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = new UiRequiredException(
+                    brokerResult.getErrorCode(),
+                    brokerResult.getErrorMessage()
+            );
+        }
+
+        if(exceptionType.equalsIgnoreCase(ServiceException.sName)){
+            Logger.warn(TAG, "Received a Service exception from Broker : "
+                    + brokerResult.getErrorCode()
+            );
+            baseException = getServiceException(brokerResult);
+        }
+
+        if(exceptionType.equalsIgnoreCase(IntuneAppProtectionPolicyRequiredException.sName)){
+            Logger.warn(TAG,
```

Comment:
May be we can move log statement outside the if blocks, and call it only once as follows:

`Logger.warn(TAG, "Received a " + exceptionType +  " from Broker : "
                    + brokerResult.getErrorCode()
            );`